pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,343 @@
+	}
+}
+
+const (
+	unlame_long = `Mark nodes as schedulable.
+`
+	unlame_example = `# Mark node "foo" as schedulable.
+$ kubectl unlame foo
+`
+)
+
+func NewCmdUnlame(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{factory: f, out: out}
+
+	return &cobra.Command{
+		Use:     "unlame NODE",
+		Short:   "Mark node as schedulable",
+		Long:    unlame_long,
+		Example: unlame_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunLame(false))
+		},
+	}
+}
+
+const (
+	drain_long = `Drain node in preparation for maintenance.
+
+The given node will be marked unschedulable to prevent new pods from arriving.
+Then all running pods will be removed from the machine.  If there are any
+unreplicated pods, no pods will be touched without --force.
+
+When you are ready to put the node back into service, use kubectl unlame.
+`
+	drain_example = `# Drain node "foo", even if there are unreplicated pods on it.
+$ kubectl drain foo --force
+
+# As above, but don't delete unreplicated pods, and use a grace period of 15 minutes.
+$ kubectl drain foo --grace-period=900
+`
+)
+
+func NewCmdDrain(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{factory: f, out: out}
+
+	cmd := &cobra.Command{
+		Use:     "drain NODE",
+		Short:   "Drain nodes in preparation for maintenance",
+		Long:    drain_long,
+		Example: drain_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunDrain(cmd))
+		},
+	}
+	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.")
+	return cmd
+}
+
+// SetupDrain populates some fields from the factory, grabs command line
+// arguments and looks up the node using Builder
+func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {
+	var err error
+	if len(args) != 1 {
+		return cmdutil.UsageError(cmd, fmt.Sprintf("USAGE: %s [flags]", cmd.Use))
+	}
+
+	if o.client, err = o.factory.Client(); err != nil {
+		return err
+	}
+
+	o.mapper, o.typer = o.factory.Object()
+
+	cmdNamespace, _, err := o.factory.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	r := o.factory.NewBuilder().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		ResourceNames("node", args[0]).
+		Do()
+
+	if err = r.Err(); err != nil {
+		return err
+	}
+
+	return r.Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+		o.nodeInfo = info
+		return nil
+	})
+}
+
+// RunDrain runs the 'drain' command
+func (o *DrainOptions) RunDrain(cmd *cobra.Command) error {
+	if err := o.RunLame(true); err != nil {
+		return err
+	}
+
+	pods, err := o.getInterestingPods()
```

Comment:
> There does not seem to be wide agreement about DaemonSet pods. They already don't obey schedulable=False, so without some other hooks the DaemonSet controller is just going to recreate the pod after we delete it.

Actually as of #17318 DaemonSet does obey unschedulable, though that's mostly just to avoid putting DaemonSet pods on the master (which now registers just like a worker node).

But I think we really should delete all pods when we drain a node.

I had asked earlier how you know a node has been drained, if there's no specific label/marker/flag/whatever for this, and IIRC you said the way you know a node has been drained is that it has no pods. So I think we really need to remove all the pods, otherwise figuring out the node has been drained will require logic that knows which kinds of pods survive draining.